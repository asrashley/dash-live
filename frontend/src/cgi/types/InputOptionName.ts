export interface InputOptionName {
    shortName: string;
    fullName: string;
    cgiName: string;
}