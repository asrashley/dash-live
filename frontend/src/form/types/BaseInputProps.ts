import type { ReadonlySignal } from "@preact/signals";
import { StaticInputProps } from "./StaticInputProps";

export type BaseInputProps = {
  name: string;
  id: string;
  type: StaticInputProps["type"];
  className: string | ReadonlySignal<string>;
  title: string;
  placeholder?: string;
  "aria-describedby": string;
  disabled: ReadonlySignal<boolean>;
  onInput: (ev: Event) => void;
};

