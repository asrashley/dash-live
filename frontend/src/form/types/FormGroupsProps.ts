import { InputFormGroup } from "./InputFormGroup";
import { FormRowMode } from "./FormRowMode";
import { SetValueFunc } from "./SetValueFunc";
import { InputProps } from "./InputProps";

export interface FormGroupsProps extends Pick<InputProps, "data" | "disabledFields"> {
  groups: InputFormGroup[];
  expand?: string;
  mode: FormRowMode;
  layout?: number[];
  setValue: SetValueFunc;
}
