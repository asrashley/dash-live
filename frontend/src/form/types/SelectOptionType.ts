export interface SelectOptionType {
    name?: string;
    title: string;
    value: string;
}
