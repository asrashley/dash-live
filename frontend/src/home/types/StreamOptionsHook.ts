import { createContext } from "preact";
import { UseStreamOptionsHook } from "../hooks/useStreamOptions";

export const StreamOptionsContext = createContext<UseStreamOptionsHook>(null);
