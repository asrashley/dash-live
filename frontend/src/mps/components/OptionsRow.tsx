import { useContext } from "preact/hooks";
import { useComputed, type ReadonlySignal } from "@preact/signals";

import { MultiPeriodModelContext } from "../../hooks/useMultiPeriodStream";
import { AppStateContext } from "../../appState";
import { FormRow } from "../../components/FormRow";
import { PrettyJson } from "../../components/PrettyJson";

export interface OptionsRowProps {
  name: string;
  canModify: ReadonlySignal<boolean>;
}

export function OptionsRow({ name, canModify }: OptionsRowProps) {
  const { model } = useContext(MultiPeriodModelContext);
  const { dialog } = useContext(AppStateContext);
  const options = useComputed(() => model.value.options ?? {});

  const openDialog = () => {
    dialog.value = {
      mpsOptions: {
        options: options.value,
        lastModified: model.value.lastModified,
        name,
      },
      backdrop: true,
    };
  };

  return (
    <FormRow name="options" label="Stream Options">
      <div className="d-flex flex-row">
        <PrettyJson className="flex-fill me-1" data={options.value} />
        {canModify.value ? (
          <button className="btn btn-primary" onClick={openDialog}>
            Options
          </button>
        ) : null}
      </div>
    </FormRow>
  );
}
