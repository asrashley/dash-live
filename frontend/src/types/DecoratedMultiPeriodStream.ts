import { MpsPeriod } from "./MpsPeriod";
import { MultiPeriodStream } from "./MultiPeriodStream";


export type DecoratedMultiPeriodStream = Omit<MultiPeriodStream, 'periods'> & {
    periods: MpsPeriod[];
    modified: boolean;
    lastModified: number;
};
