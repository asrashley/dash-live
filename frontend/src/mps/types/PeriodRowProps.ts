import { MpsPeriod } from "../../types/MpsPeriod";

export interface PeriodRowProps {
  className?: string;
  index: number;
  item: MpsPeriod;
}
