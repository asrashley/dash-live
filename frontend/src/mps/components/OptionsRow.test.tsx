import { beforeEach, describe, expect, test } from "vitest";
import { mock } from "vitest-mock-extended";
import { signal } from "@preact/signals";

import { renderWithProviders } from "../../test/renderWithProviders";
import { OptionsRow } from "./OptionsRow";
import {
  decorateMultiPeriodStream,
  MpsModelValidationErrors,
  MultiPeriodModelContext,
  UseMultiPeriodStreamHook,
} from "../../hooks/useMultiPeriodStream";
import { DecoratedMultiPeriodStream } from "../../types/DecoratedMultiPeriodStream";

import { model as mpsStream } from "../../test/fixtures/multi-period-streams/demo.json";
import { fireEvent } from "@testing-library/preact";

describe("OptionsRow component", () => {
  const canModify = signal<boolean>(false);
  const model = signal<DecoratedMultiPeriodStream>();
  const multiPeriodStreamHook = mock<UseMultiPeriodStreamHook>({
    loaded: signal<string | undefined>(),
    modified: signal<boolean>(false),
    errors: signal<MpsModelValidationErrors>({}),
    isValid: signal<boolean>(true),
    model,
  });

  beforeEach(() => {
    model.value = decorateMultiPeriodStream(mpsStream);
    canModify.value = true;
  });

  test("matches snapshot", () => {
    const { asFragment } = renderWithProviders(
      <MultiPeriodModelContext.Provider value={multiPeriodStreamHook}>
        <OptionsRow name="rowtest" canModify={canModify} />
      </MultiPeriodModelContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("does not show options button if modification not allowed", () => {
    canModify.value = false;
    const { queryByText } = renderWithProviders(
      <MultiPeriodModelContext.Provider value={multiPeriodStreamHook}>
        <OptionsRow name="rowtest" canModify={canModify} />
      </MultiPeriodModelContext.Provider>
    );
    expect(queryByText("Options")).toBeNull();
  });

  test("can open options dialog", () => {
    const { getByText, state } = renderWithProviders(
      <MultiPeriodModelContext.Provider value={multiPeriodStreamHook}>
        <OptionsRow name="rowtest" canModify={canModify} />
      </MultiPeriodModelContext.Provider>
    );
    const btn = getByText("Options") as HTMLButtonElement;
    fireEvent.click(btn);
    expect(state.dialog.value).toEqual({
      backdrop: true,
      mpsOptions: {
        lastModified: 0,
        name: "rowtest",
        options: mpsStream.options,
      },
    });
  });
});
