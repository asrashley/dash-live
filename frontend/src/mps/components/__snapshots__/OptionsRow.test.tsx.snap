// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OptionsRow component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="row mb-2 form-group "
  >
    <label
      class="col-2 col-form-label"
      for="model-options"
      id="label-options"
    >
      Stream Options:
    </label>
    <div
      class="col-10"
    >
      <div
        class="d-flex flex-row"
      >
        <div
          class="json bg-secondary-subtle border border-dark-subtle flex-fill me-1"
        >
          <span
            class="key"
          >
            ntps
          </span>
          : 
          <span
            class="value"
          >
            "europe-ntp"
          </span>
          , 
          <span
            class="key"
          >
            st
          </span>
          : 
          <span
            class="value"
          >
            true
          </span>
        </div>
        <button
          class="btn btn-primary"
        >
          Options
        </button>
      </div>
    </div>
    <div
      class="invalid-feedback"
      style="display: none;"
    />
  </div>
</DocumentFragment>
`;
